leap/rk/__init__.py

```diff
@@ -203,166 +258,157 @@ def generate_butcher(self, stage_coeff_set_names, stage_coeff_sets, rhs_funcs,
+                        rhs_expr = rhs_funcs[name](
+                                t=t + c*dt, **{comp: state_est})
+
+                        cb(my_rhs, rhs_expr)
+                        make_known(my_rhs)
+
+                # {{{ emit solve if we have any unknowns/equations
+
+                if unknowns and len(unknowns) == len(equations):
+                    from leap.implicit import generate_solve
+                    generate_solve(cb, unknowns, equations, rhs_var_to_unknown,
+                                    self.state)
+                elif not unknowns:
+                    # we have an explicit Runge-Kutta method
+                    pass
+                elif len(unknowns) > len(equations):
+                    raise ValueError("Runge-Kutta implicit timestep has more "
+                            "unknowns than equations")
+                elif len(unknowns) < len(equations):
+                    raise ValueError("Runge-Kutta implicit timestep has more "
+                            "equations than unknowns")
+
+                del equations[:]
+                knowns.update(unknowns)
+                unknowns.clear()
+
+                # }}}
+
+        # Compute solution estimates.
+        estimate_vars = [
+                cb.fresh_var("est_"+name)
+                for name in estimate_coeff_set_names]
+
+        for iest, name in enumerate(estimate_coeff_set_names):
+            out_coeffs = estimate_coeff_sets[name]
+
+            if (
+                    last_state_est_var_valid
+                    and  # noqa: W504
+                    _is_last_stage_same_as_output(self.c,
+                        stage_coeff_sets, out_coeffs)):
+                state_est = last_state_est_var
+
+            else:
+                state_increment = 0
+                for src_name in stage_coeff_set_names:
+                    state_increment += sum(
+                                coeff * stage_rhs_vars[src_name][src_istage]
+                                for src_istage, coeff in enumerate(out_coeffs))
+
+                state_est = self.state + dt*state_increment
+
+                if self.state_filter is not None:
+                    state_est = self.state_filter(state_est)
+
+            cb(
+                estimate_vars[iest],
+                state_est)
 
         cb_primary = cb
 
-        # }}}
+        return cb_primary, stage_rhs_vars, estimate_vars
 
-        return DAGCode(
-                phases={
-                    "initial": cb_init.as_execution_phase(next_phase="primary"),
-                    "primary": cb_primary.as_execution_phase(next_phase="primary")
-                    },
-                initial_phase="initial")
+    # }}}
+
+    # {{{ update state and time
+
+    def generate_butcher_finish(self, cb, stage_coeff_set_names,
+            stage_coeff_sets, rhs_funcs, estimate_coeff_set_names,
+            estimate_coeff_sets, stage_rhs_vars, estimate_vars):
```

Comment:
Explain the circumstances in which "finish" is invoked.

Reply:
The reason that "finish" exists separately from "primary" essentially relates to the use of these functions in bootstrapping the multistep methods. This "finish" is invoked immediately after "primary" by regularly generated RK methods, because one of its primary responsibilities is yielding state and incrementing the time, but when bootstrapping a multistep method, this "finish" is not called, because other pieces of code are responsible for doing those things (not to mention there are instances in which other things need to occur in the "finish" of a bootstrap step for a multistep method pertaining to history updates).